In [ ]:
#importing Deopendencies
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [ ]:
#loading the dataset into Pandas DataFrame
credit_card = pd.read_csv('/content/creditcard.csv')

In [ ]:
#first five rows of the dataset
credit_card.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [ ]:
#last five rows of the dataset
credit_card.tail()

Time         V1         V2  ...       V28  Amount  Class
284802  172786.0 -11.881118  10.071785  ...  0.823731    0.77      0
284803  172787.0  -0.732789  -0.055080  ... -0.053527   24.79      0
284804  172788.0   1.919565  -0.301254  ... -0.026561   67.88      0
284805  172788.0  -0.240440   0.530483  ...  0.104533   10.00      0
284806  172792.0  -0.533413  -0.189733  ...  0.013649  217.00      0

[5 rows x 31 columns]

In [ ]:
#dataset information
credit_card.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [ ]:
#checking number of missing value in each colum
credit_card.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
#checking the shape of the dataset
credit_card.shape

(284807, 31)

In [ ]:
#droping the Null values/missing values
credit_card_data = credit_card.dropna(how='any')

In [ ]:
credit_card_data.shape

(284807, 31)

In [ ]:
#checking null values agin
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
#distribudtion of legit and fraudulent transactions
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

Dataset is highly imbalanced


0 ----> normal transactions


1 ----> Fraudulent transactions

In [ ]:
#separating the data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [ ]:
#statistical measures of the data
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [ ]:
#compare the values for both transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      94838.202258  0.008258 -0.006271  ... -0.000295 -0.000131   88.291022
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

Under-Sampling

Build a sample dataset containing similar distribution of normal and fraudulent transactions

Number of Fradulent transactions ---> 492

Number of Normal transictions ---> 284315

In [ ]:
legit_sample = legit.sample(n=492)

In [ ]:
#Concatennating two dataframes
new_dataset = pd.concat([legit_sample, fraud],axis=0)

In [ ]:
new_dataset.head()

Time        V1        V2  ...       V28  Amount  Class
230495  146352.0  2.058866  0.088904  ... -0.044753    1.79      0
153172   98101.0  1.999767 -0.161425  ... -0.011604    2.00      0
59826    49022.0 -1.943779 -1.829324  ... -0.046284   57.25      0
142587   84809.0  1.146441  1.182742  ...  0.084914    8.99      0
138797   82847.0 -0.371215 -0.217002  ...  0.145483   77.65      0

[5 rows x 31 columns]

In [ ]:
new_dataset.tail()

Time        V1        V2  ...       V28  Amount  Class
279863  169142.0 -1.927883  1.125653  ...  0.147968  390.00      1
280143  169347.0  1.378559  1.289381  ...  0.186637    0.76      1
280149  169351.0 -0.676143  1.126366  ...  0.194361   77.89      1
281144  169966.0 -3.113832  0.585864  ... -0.253700  245.00      1
281674  170348.0  1.991976  0.158476  ... -0.015309   42.53      1

[5 rows x 31 columns]

In [ ]:
new_dataset['Class'].value_counts()

1    492
0    492
Name: Class, dtype: int64

In [ ]:
new_dataset.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      97792.184959  0.009774  0.069072  ...  0.010728  0.003879   82.023902
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

slpitting the data into Features and Targets

In [ ]:
X = new_dataset.drop(columns='Class', axis =1)
Y = new_dataset['Class']

In [ ]:
print(X)

            Time        V1        V2  ...       V27       V28  Amount
230495  146352.0  2.058866  0.088904  ... -0.067797 -0.044753    1.79
153172   98101.0  1.999767 -0.161425  ... -0.022623 -0.011604    2.00
59826    49022.0 -1.943779 -1.829324  ... -0.360357 -0.046284   57.25
142587   84809.0  1.146441  1.182742  ...  0.051634  0.084914    8.99
138797   82847.0 -0.371215 -0.217002  ...  0.104342  0.145483   77.65
...          ...       ...       ...  ...       ...       ...     ...
279863  169142.0 -1.927883  1.125653  ...  0.292680  0.147968  390.00
280143  169347.0  1.378559  1.289381  ...  0.389152  0.186637    0.76
280149  169351.0 -0.676143  1.126366  ...  0.385107  0.194361   77.89
281144  169966.0 -3.113832  0.585864  ...  0.884876 -0.253700  245.00
281674  170348.0  1.991976  0.158476  ...  0.002988 -0.015309   42.53

[984 rows x 30 columns]


In [ ]:
print(Y)

230495    0
153172    0
59826     0
142587    0
138797    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [ ]:
#spilting into training and testing 
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

In [ ]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


Model Training

In [ ]:
#Logistic Regression
model = LogisticRegression(multi_class='auto',solver='lbfgs',tol=0.0001, max_iter=200)

In [ ]:
#training the model
model.fit(X_train, Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=200,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

Evaluation

In [ ]:
#Accuracy Score on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [ ]:
print("Accuracy on training data : ", training_data_accuracy)

Accuracy on training data :  0.9453621346886912


In [ ]:
#Accuracy Score on Testing data
X_test_prediction = model.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [ ]:
print("Accuracy on testing data : ", testing_data_accuracy)

Accuracy on testing data :  0.9441624365482234
